src/hotspot/share/gc/shenandoah/shenandoahMark.inline.hpp

```diff
@@ -285,23 +285,20 @@ inline void ShenandoahMark::mark_through_ref(T *p, ShenandoahObjToScanQueue* q,
     ShenandoahHeap* heap = ShenandoahHeap::heap();
     shenandoah_assert_not_forwarded(p, obj);
     shenandoah_assert_not_in_cset_except(p, obj, heap->cancelled_gc());
-    if (in_generation<GENERATION>(obj)) {
+    if (in_generation<GENERATION>(heap, obj)) {
       mark_ref(q, mark_context, weak, obj);
       shenandoah_assert_marked(p, obj);
-      // TODO: This is v-call on very hot path, can we sense the same from GENERATION?
-      if (heap->mode()->is_generational()) {
-        // TODO: As implemented herein, GLOBAL collections reconstruct the card table during GLOBAL concurrent
-        // marking. Note that the card table is cleaned at init_mark time so it needs to be reconstructed to support
-        // future young-gen collections.  It might be better to reconstruct card table in
-        // ShenandoahHeapRegion::global_oop_iterate_and_fill_dead.  We could either mark all live memory as dirty, or could
-        // use the GLOBAL update-refs scanning of pointers to determine precisely which cards to flag as dirty.
-        if (GENERATION == YOUNG && heap->is_in_old(p)) {
-          // Mark card as dirty because remembered set scanning still finds interesting pointer.
-          heap->mark_card_as_dirty((HeapWord*)p);
-        } else if (GENERATION == GLOBAL && heap->is_in_old(p) && heap->is_in_young(obj)) {
-          // Mark card as dirty because GLOBAL marking finds interesting pointer.
-          heap->mark_card_as_dirty((HeapWord*)p);
-        }
+      // TODO: As implemented herein, GLOBAL_GEN collections reconstruct the card table during GLOBAL_GEN concurrent
+      // marking. Note that the card table is cleaned at init_mark time so it needs to be reconstructed to support
+      // future young-gen collections.  It might be better to reconstruct card table in
+      // ShenandoahHeapRegion::global_oop_iterate_and_fill_dead.  We could either mark all live memory as dirty, or could
+      // use the GLOBAL update-refs scanning of pointers to determine precisely which cards to flag as dirty.
+      if (GENERATION == YOUNG && heap->is_in_old(p)) {
```

Comment:
May merit a comment to remind the reader that lines 296-302 disappear if GENERATION==GLOBAL_NON_GEN

Reply:
Added a blurb!